dashboard/app/assets/javascripts/header.js

```diff
@@ -445,16 +445,28 @@ function mergedActivityCssClass(a, b) {
 }
 
 function populateProgress(scriptName) {
-  var scriptProgress = dashboard.clientState.allLevelsProgress()[scriptName] || {};
+  var clientProgress = dashboard.clientState.allLevelsProgress()[scriptName] || {};
+  $.ajax('/api/user_progress/' + scriptName).done(function (data) {
+    if (!data || !data.levels) {
+      return;
+    }
+    var serverProgress = data.levels;
 
-  // Aggregate progress from server and client
-  $.each(scriptProgress, function (level_id, result) {
-    var level_link = $('#level-' + level_id);
-    var status = mergedActivityCssClass(result, level_link.data('result'));
+    // Aggregate progress from server and client
+    var keys = Object.keys(clientProgress).concat(Object.keys(serverProgress));
+    var uniqueKeys = {};
+    keys.forEach(function (key) {
+      uniqueKeys[key] = true;
+    });
 
-    if (!level_link.hasClass(status)) {
-      // Clear the existing class and replace
-      level_link.attr('class', 'level_link ' + status);
-    }
+    Object.keys(uniqueKeys).forEach(function (levelId) {
+      var level_link = $('#level-' + levelId);
+      var status = mergedActivityCssClass(clientProgress[levelId], serverProgress[levelId].result);
```

Comment:
I think we always set it to an object here: https://github.com/code-dot-org/code-dot-org/blob/b90f5985a1c41b75ca50bbf0c67631e06a43df58/dashboard/app/helpers/users_helper.rb#L42

I can defend against the null here, but I almost prefer having the API only set the key if it has progress data for that level.